ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -373,6 +391,57 @@ void fillChunks() {
   }
 
   /**
+   * Submits encrypt job for the given {@link PutChunk} and processes the callback for the same
+   * @param putChunk the {@link PutChunk} that needs to be encrypted
+   * @param encryptContent {@code true} if content needs to be encrypted. {@code false} otherwise
+   */
+  private void encryptChunk(PutChunk putChunk, boolean encryptContent) {
+    try {
+      logger.trace("Chunk at index {} state moves to {}", putChunk.chunkIndex, ChunkState.Encrypting);
+      putChunk.state = ChunkState.Encrypting;
+      if (encryptContent) {
+        putChunk.buf.flip();
+        putChunk.chunkBlobSize = putChunk.buf.remaining();
+      }
+      logger.trace("Submitting encrypt job for chunk at index " + putChunk.chunkIndex);
+      exec.submitJob(new EncryptJob(passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+          encryptContent ? putChunk.buf : null, ByteBuffer.wrap(putChunk.chunkUserMetadata), kms.getRandomKey(),
+          cryptoService, kms, (EncryptJob.EncryptJobResult result, Exception exception) -> {
+        logger.trace("Processing encrypt job callback for chunk at index " + putChunk.chunkIndex);
+        if (exception == null && !isOperationComplete()) {
+          if (encryptContent) {
+            putChunk.buf = result.getEncryptedBlobContent();
+          }
+          putChunk.encryptedPerBlobKey = result.getEncryptedKey();
+          putChunk.chunkUserMetadata = result.getEncryptedUserMetadata().array();
+          logger.trace("Completing encrypt job result for chunk at index " + putChunk.chunkIndex);
+          putChunk.onFillComplete(true);
+          // todo: add metrics to track encryption time
+        } else {
+          if (getOperationException() == null) {
+            logger.trace("Setting exception {} from encrypt of chunk at index {} ", exception.getCause(),
+                putChunk.chunkIndex);
+            setOperationExceptionAndComplete(new RouterException(
+                "Exception thrown on encrypting the content for chunk at index " + putChunk.chunkIndex,
+                RouterErrorCode.UnexpectedInternalError));
+          } else {
+            logger.trace(
+                "Ignoring exception from encrypt job for chunk at index {} as operation exception {} is set already",
+                putChunk.chunkIndex, getOperationException().getCause());
+          }
+        }
+        routerCallback.onPollReady();
```

Comment:
do you have to call this regardless of whether there is an exception or not? If yes, why don't you do in the `catch`?

Reply:
catch block is for kms.getRandomKey(). This is not in the callback processing. This will be thrown even before job is submitted. So, its in regular path. In regular execution path, routerCallback.onPollReady() will be invoked after encryptChunk()